### ISSUE TITLE
Building failed with "incompatible pointer type" in os_dep/linux/ioctl_cfg80211.c

### ISSUE THREAD

**Issue by FomalhautWeisszwerg** [FomalhautWeisszwerg/rtl8822bu#2]
Found on b281b12ffcdc7a7c0a8fc2e1e9f6eaaa33a8402d

OS

```
$ uname -a
Linux debian-server 4.12.0-1-amd64 #1 SMP Debian 4.12.6-1 (2017-08-12) x86_64 GNU/Linux
```

GCC

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 7.1.0-13' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 7.1.0 (Debian 7.1.0-13)
```

Log

```
$ git show
commit b281b12ffcdc7a7c0a8fc2e1e9f6eaaa33a8402d (HEAD -> chase_to_linux-4.12, origin/chase_to_linux-4.12)
Author: Fomalhaut Weisszwerg <weisszwerg@gmail.com>
Date:   Thu Aug 17 19:46:33 2017 +0900

    Fix: implicit declaration of function 'allow_signal'.

    'allow_signal' requires explicit including of "linux/sched/signal.h" from Linux 4.11.

diff --git a/include/osdep_service.h b/include/osdep_service.h
index 59b01e3..c846548 100644
--- a/include/osdep_service.h
+++ b/include/osdep_service.h
@@ -39,6 +39,7 @@

 #ifdef PLATFORM_LINUX
        #include <osdep_service_linux.h>
+       #include <linux/sched/signal.h>
 #endif

 #ifdef PLATFORM_OS_XP
$ ls
clean  hal          include  Makefile  platform   RTL8822Bfw_NIC.bin  runwpa
core   ifcfg-wlan0  Kconfig  os_dep    README.md  rtl8822b.mk         wlan0dhcp
$ make
make ARCH=x86_64 CROSS_COMPILE= -C /lib/modules/4.12.0-1-amd64/build M=/home/working/rtl8822bu modules
make[1]: Entering directory '/usr/src/linux-headers-4.12.0-1-amd64'
  CC [M]  /home/working/rtl8822bu/core/rtw_cmd.o
  CC [M]  /home/working/rtl8822bu/core/rtw_security.o
  CC [M]  /home/working/rtl8822bu/core/rtw_debug.o
  CC [M]  /home/working/rtl8822bu/core/rtw_io.o
  CC [M]  /home/working/rtl8822bu/core/rtw_ioctl_query.o
  CC [M]  /home/working/rtl8822bu/core/rtw_ioctl_set.o
  CC [M]  /home/working/rtl8822bu/core/rtw_ieee80211.o
  CC [M]  /home/working/rtl8822bu/core/rtw_mlme.o
  CC [M]  /home/working/rtl8822bu/core/rtw_mlme_ext.o
  CC [M]  /home/working/rtl8822bu/core/rtw_mi.o
  CC [M]  /home/working/rtl8822bu/core/rtw_wlan_util.o
  CC [M]  /home/working/rtl8822bu/core/rtw_vht.o
  CC [M]  /home/working/rtl8822bu/core/rtw_pwrctrl.o
  CC [M]  /home/working/rtl8822bu/core/rtw_rf.o
  CC [M]  /home/working/rtl8822bu/core/rtw_recv.o
  CC [M]  /home/working/rtl8822bu/core/rtw_sta_mgt.o
  CC [M]  /home/working/rtl8822bu/core/rtw_ap.o
  CC [M]  /home/working/rtl8822bu/core/rtw_xmit.o
  CC [M]  /home/working/rtl8822bu/core/rtw_p2p.o
  CC [M]  /home/working/rtl8822bu/core/rtw_tdls.o
  CC [M]  /home/working/rtl8822bu/core/rtw_br_ext.o
  CC [M]  /home/working/rtl8822bu/core/rtw_iol.o
  CC [M]  /home/working/rtl8822bu/core/rtw_sreset.o
  CC [M]  /home/working/rtl8822bu/core/rtw_btcoex.o
  CC [M]  /home/working/rtl8822bu/core/rtw_beamforming.o
  CC [M]  /home/working/rtl8822bu/core/rtw_odm.o
  CC [M]  /home/working/rtl8822bu/core/efuse/rtw_efuse.o
  CC [M]  /home/working/rtl8822bu/os_dep/osdep_service.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/os_intfs.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/usb_intf.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/usb_ops_linux.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/ioctl_linux.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/xmit_linux.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/mlme_linux.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/recv_linux.o
  CC [M]  /home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.o
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c: In function ‘rtw_cfg80211_indicate_connect’:
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:753:21: error: passing argument 2 of ‘cfg80211_roamed’ from incompatible pointer type [-Werror=incompatible-pointer-types]
                   , notify_channel
                     ^~~~~~~~~~~~~~
In file included from /home/working/rtl8822bu/include/osdep_service_linux.h:87:0,
                 from /home/working/rtl8822bu/include/osdep_service.h:41,
                 from /home/working/rtl8822bu/include/drv_types.h:32,
                 from /home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:22:
/usr/src/linux-headers-4.12.0-1-common/include/net/cfg80211.h:5435:6: note: expected ‘struct cfg80211_roam_info *’ but argument is of type ‘struct ieee80211_channel *’
 void cfg80211_roamed(struct net_device *dev, struct cfg80211_roam_info *info,
      ^~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:755:21: warning: passing argument 3 of ‘cfg80211_roamed’ makes integer from pointer without a cast [-Wint-conversion]
                   , cur_network->network.MacAddress
                     ^~~~~~~~~~~
In file included from /home/working/rtl8822bu/include/osdep_service_linux.h:87:0,
                 from /home/working/rtl8822bu/include/osdep_service.h:41,
                 from /home/working/rtl8822bu/include/drv_types.h:32,
                 from /home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:22:
/usr/src/linux-headers-4.12.0-1-common/include/net/cfg80211.h:5435:6: note: expected ‘gfp_t {aka unsigned int}’ but argument is of type ‘unsigned char *’
 void cfg80211_roamed(struct net_device *dev, struct cfg80211_roam_info *info,
      ^~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:751:3: error: too many arguments to function ‘cfg80211_roamed’
   cfg80211_roamed(padapter->pnetdev
   ^~~~~~~~~~~~~~~
In file included from /home/working/rtl8822bu/include/osdep_service_linux.h:87:0,
                 from /home/working/rtl8822bu/include/osdep_service.h:41,
                 from /home/working/rtl8822bu/include/drv_types.h:32,
                 from /home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:22:
/usr/src/linux-headers-4.12.0-1-common/include/net/cfg80211.h:5435:6: note: declared here
 void cfg80211_roamed(struct net_device *dev, struct cfg80211_roam_info *info,
      ^~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c: In function ‘rtw_cfg80211_add_monitor_if’:
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:3598:10: error: ‘struct net_device’ has no member named ‘destructor’; did you mean ‘priv_destructor’?
  mon_ndev->destructor = rtw_ndev_destructor;
          ^~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c: In function ‘rtw_cfg80211_preinit_wiphy’:
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6097:18: error: ‘WIPHY_FLAG_SUPPORTS_SCHED_SCAN’ undeclared (first use in this function)
  wiphy->flags |= WIPHY_FLAG_SUPPORTS_SCHED_SCAN;
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6097:18: note: each undeclared identifier is reported only once for each function it appears in
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c: At top level:
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6130:25: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .change_virtual_intf = cfg80211_rtw_change_iface,
                         ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6130:25: note: (near initialization for ‘rtw_cfg80211_ops.change_virtual_intf’)
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6153:22: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .add_virtual_intf = cfg80211_rtw_add_virtual_intf,
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.c:6153:22: note: (near initialization for ‘rtw_cfg80211_ops.add_virtual_intf’)
cc1: some warnings being treated as errors
/usr/src/linux-headers-4.12.0-1-common/scripts/Makefile.build:307: recipe for target '/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.o' failed
make[4]: *** [/home/working/rtl8822bu/os_dep/linux/ioctl_cfg80211.o] Error 1
/usr/src/linux-headers-4.12.0-1-common/Makefile:1532: recipe for target '_module_/home/working/rtl8822bu' failed
make[3]: *** [_module_/home/working/rtl8822bu] Error 2
Makefile:152: recipe for target 'sub-make' failed
make[2]: *** [sub-make] Error 2
Makefile:8: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.12.0-1-amd64'
Makefile:1318: recipe for target 'modules' failed
make: *** [modules] Error 2
```

**Comment by FomalhautWeisszwerg** [comment]
The function `cfg80211_roamed()` and `cfg80211_roamed_bss()` ware unified by commit [29ce6ec] in Linux Kernel.
See: https://github.com/torvalds/linux/commit/29ce6ecbb83c9185d76e3a7c340c9702d2a54961#diff-6788d961d43074bcd7966d231861f6b8

**Comment by FomalhautWeisszwerg** [comment]
Fixed at [1cdec984eab76df3423148ce5c59c1a9b59cca58] but the other problems remain.
